middle_end/flambda2/to_cmm/to_cmm_env.ml

```diff
@@ -626,16 +596,115 @@ let split_and_inline env res var binding =
           bindings = Variable.Map.add var (Binding split_binding) env.bindings
         }
     in
-    let env =
+    let env, res =
       List.fold_left
-        (fun env new_binding ->
+        (fun (env, res) new_binding ->
           let flambda_var = Variable.create "to_cmm_tmp" in
-          { env with
-            bindings = Variable.Map.add flambda_var new_binding env.bindings
-          })
-        env new_bindings
+          add_binding_to_env env res flambda_var new_binding)
+        (env, res) new_bindings
+    in
+    env, res, split_binding
+
+let bind_variable_with_decision (type a) ?extra env res var ~inline
+    ~(defining_expr : a bound_expr) ~effects_and_coeffects_of_defining_expr:effs
+    =
+  let effs =
+    let classification =
+      To_cmm_effects.classify_by_effects_and_coeffects effs
     in
-    will_inline_complex env res split_binding
+    match[@ocaml.warning "-4"] (inline : a inline), classification with
+    | (Must_inline_once | Must_inline_and_duplicate), Generative_immutable -> (
+      (* Effects (including generative immutbale effects) can severly limit the
+         inlining of bindings (due to the
+         [~consider_inlining_effectful_expressions]). See CR above for a more
+         lengthy explanation.
+
+         Therefore, we try and "forget" generative effects here. This is
+         reasonable since the only reason that `Generative_immutable' effects
+         are considered effects is to prevent allocations from being moved
+         across a `Gc` function call (and the main point of that is to not
+         invalidate allocation tests). We consider that the Delay placement on a
+         allocating primitive explicitly allows re-ordering past `Gc` function
+         calls, and therefore it is correct to ignore the generative effects; we
+         also do that when computing the effects after splitting a binding (see
+         [split_complex_binding]).
+
+         Note that in some cases, there might be situations where an allocation
+         with Strict placement might be inlined inside a Delay allocation, and
+         end up moved across a `Gc` function call beause of this. Such cases are
+         currently impossible since only [Box_number] and [Project_value_slot]
+         have a Delay placement, but may appaer in the future if/when some more
```

Comment:
```suggestion
         have a Delay placement, but may appear in the future if/when some more
```
Also note that `Project_value_slot` can have arguments with `Generative_immutable` effects (the closures), so I think we can hit this issue in practice.

Reply:
Yes, but the set of closure allocation should have a 'never inline' decision, so that should not happen

Reply:
Why would you expect it to be "never inline" ? If it's only used once, for the projection, it should be `May_inline_once`. Unless I'm missing something ?

Reply:
Cf https://github.com/ocaml-flambda/flambda-backend/blob/2a3b9de340043988fb742dc605777182c09a484e/middle_end/flambda2/to_cmm/to_cmm_set_of_closures.ml#L523

Basically, we currently in prevent inlining of sets of closures, because as far as I recall, there's no optimization that can make use of it.

Reply:
In any case, the typo is now fixed.